### ISSUE TITLE
[flang] Incorrect diagnostic on defined operator

### ISSUE THREAD

**Issue by DanielCChen** [llvm/llvm-project#148678]
Consider the following code:
```
module m
  type dt
    integer i
  contains
    procedure, pass(x) :: add => addmp
    generic, private :: OPERATOR(+) => add
  end type

  interface OPERATOR(+)
    module procedure addmp
  end interface
contains
  type(dt) function addmp(y,x)
    class(dt), intent(in) :: x
    integer, intent(in) :: y
    addmp%i = x%i + y + 1
  end function

end module
```

Flang complains:
```
error: Semantic errors in genop03.f
./genop03.f:8:25: error: Generic 'OPERATOR(+)' may not have specific procedures 'addmp' and 'dt%add' as their interfaces are not distinguishable
      generic, private :: OPERATOR(+) => add
                          ^^^^^^^^^^^
./genop03.f:15:21: Declaration of 'addmp'
    type(dt) function addmp(y,x)
                      ^^^^^
./genop03.f:7:27: Declaration of 'add'
      procedure, pass(x) :: add => addmp
                            ^^^
./genop03.f:15:21: Procedure 'add' of type 'dt' is bound to 'addmp'
    type(dt) function addmp(y,x)
                      ^^^^^
```

All ifort, gfortran and XLF compiles the code successfully.

**Comment by klausler** [comment]
I think that the error is valid, since it is possible that the binding could be overridden in an extended type in the same module.

**Comment by klausler** [comment]
Further, 10.1.6.1 p6 (2) contains the word "either", so an ambiguous definition like this is not admissible.

**Comment by DanielCChen** [comment]
The intention of the test case is to provide an alternative way to do the defined operation. Here is another example code that better illustrates that.
```
module m
  type base
    integer i
  contains
    private
    procedure :: powerint
    generic, private :: operator(**) => powerint
  end type

  interface operator(**)
    module procedure powerint
  end interface

contains
  type(base) function powerint(x, y)
    class(base), intent(in) :: x
    integer, intent(in) :: y
    powerint%i = x%i + y
  end function

end module

use m
type(base) :: a = base(3)
type(base) b

b = a ** 1                                 !! Need interface to work.
if (b%i /= 4) then
  print *, b%i
  ERROR STOP 1
endif
end
```

In this example, `powerint` is private, so the line `b = a **1` requires the interface block to work.

**Comment by klausler** [comment]
That technique won't work if the interface block is in the same module as the derived type definition.  But you could move the interface block into an intervening module, like

```
module m2
  use m
  interface operator(**)
    procedure powerint
  end interface
end module
```

and then use that module.

Note that at least one other compiler also detects and enforces this error.
